File: ambeth/Ambeth.Test/ambeth/xml/oriwrapper/OriWrapperReadTest.cs
using System;
using System.Collections;
using System.Collections.Generic;
using De.Osthus.Ambeth.Config;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Test.Model;
using De.Osthus.Ambeth.Testutil;
using De.Osthus.Ambeth.Transfer;
using De.Osthus.Ambeth.Util;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace De.Osthus.Ambeth.Xml.Test
{
    [TestClass]
    [TestProperties(File = "ambeth/xml/oriwrapper/OriWrapperTestData.properties")]
    [TestProperties(Name = ServiceConfigurationConstants.MappingFile, Value = "ambeth/xml/oriwrapper/orm.xml")]
    [TestModule(typeof(BootstrapScannerModule), typeof(OriWrapperTestModule), typeof(XmlModule))]
    public class OriWrapperReadTest : AbstractInformationBusTest
    {
        [Autowired(XmlModule.CYCLIC_XML_HANDLER)]
        public ICyclicXmlHandler CyclicXmlHandler { protected get; set; }

        [Autowired]
        public OriWrapperTestBed OriWrapperTestBed { protected get; set; }

        public override void AfterPropertiesSet()
        {
            base.AfterPropertiesSet();

            OriWrapperTestBed.Init();
        }

        /// <summary>
        /// Workaround to get all needed assemblies known by the AssemblyHelper.
        /// </summary>
        /// <param name="context"></param>
        [AssemblyInitialize]
        public static new void RegisterAssemblies(Microsoft.VisualStudio.TestTools.UnitTesting.TestContext context)
        {
            AbstractIocTest.RegisterAssemblies(context);

            AssemblyHelper.RegisterAssemblyFromType(typeof(Material));
        }

        [TestMethod]
        public void ReadSimpleEntity()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getSimpleEntityTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertEquals(testData.obj, obj);
        }

        [TestMethod]
        public void ReadEntityWithRelation()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getEntityWithRelationTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertEquals(testData.obj, obj);
        }

        [TestMethod]
        public void ReadMixedArray()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getMixedArrayTestData();
            object obj = CyclicXmlHandler.Read(testData.xml);
            AreArraysEqual((object[])testData.obj, (object[])obj);
        }

        [TestMethod]
        public void ReadMixedList()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getMixedListTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            AreCollectionsEqual((ICollection)testData.obj, (ICollection)obj);
        }

        [TestMethod]
        public void ReadMixedLinkedSet()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getMixedLinkedSetTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            AreSetsEqual((ISet<Object>)testData.obj, (ISet<Object>)obj);
        }

        [TestMethod]
        public void ReadServiceDescription()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getServiceDescriptionTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            De.Osthus.Ambeth.Transfer.Assert.AssertEquals((ServiceDescription)testData.obj, (ServiceDescription)obj);
        }

        [TestMethod]
        public void ReadCreatedEntity()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedEntityTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertTrue(obj is Material);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadCreatedEntityJavaXML()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedEntityTestData();
            Object obj = CyclicXmlHandler.Read(testData.xmlJava);
            Assert.AssertTrue(obj is Material);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadCreatedChildEntity()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedChildEntityTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertTrue(obj is Material);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadCreatedChildEntity2()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedChildEntityTestData2();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertTrue(obj is Material);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadCreatedChildEntityJavaXML()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedChildEntityTestData();
            Object obj = CyclicXmlHandler.Read(testData.xmlJava);
            Assert.AssertTrue(obj is Material);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadCreatedParentAndChildEntities()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedParentAndChildEntitiesTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertTrue(obj is Material);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadCreatedParentAndChildEntities2()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedParentAndChildEntitiesTestData2();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertTrue(obj is Material);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadCreatedParentAndChildEntitiesJavaXML()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedParentAndChildEntitiesTestData();
            Object obj = CyclicXmlHandler.Read(testData.xmlJava);
            Assert.AssertTrue(obj is Material);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadCreatedParentAndChildEntitiesInList()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedParentAndChildEntitiesInListTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertTrue(obj is IList);
            IList<Object> actuals = (IList<Object>)obj;
            AssertEquals((Material)((IList<Object>)testData.obj)[0], (Material)actuals[0]);
        }

        [TestMethod]
        public void ReadCreatedParentAndChildEntitiesInList2()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedParentAndChildEntitiesInListTestData2();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertTrue(obj is IList);
            IList<Object> actuals = (IList<Object>)obj;
            AssertEquals((Material)((IList<Object>)testData.obj)[0], (Material)actuals[0]);
        }

        [TestMethod]
        public void ReadCreatedParentAndChildEntitiesInListJavaXML()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedParentAndChildEntitiesInListTestData();
            Object obj = CyclicXmlHandler.Read(testData.xmlJava);
            Assert.AssertTrue(obj is IList);
            IList<Object> actuals = (IList<Object>)obj;
            AssertEquals((Material)((IList<Object>)testData.obj)[0], (Material)actuals[0]);
        }

        [TestMethod]
        public void ReadMultipleCreatedEntities()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getMultipleCreatedEntitiesTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            List<Object> expecteds = (List<Object>)testData.obj;
            List<Object> actuals = (List<Object>)obj;
            Assert.AssertEquals(expecteds.Count, actuals.Count);
            for (int i = 0; i < expecteds.Count; i++)
            {
                Material expected = expecteds[i] as Material;
                Material actual = actuals[i] as Material;
                AssertEquals(expected, actual);
            }
        }

        [TestMethod]
        public void ReadMultipleCreatedEntities2()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getMultipleCreatedEntitiesTestData2();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            List<Object> expecteds = (List<Object>)testData.obj;
            List<Object> actuals = (List<Object>)obj;
            Assert.AssertEquals(expecteds.Count, actuals.Count);
            for (int i = 0; i < expecteds.Count; i++)
            {
                Material expected = expecteds[i] as Material;
                Material actual = actuals[i] as Material;
                AssertEquals(expected, actual);
            }
        }

        [TestMethod]
        public void ReadMultipleCreatedEntitiesJavaXML()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getMultipleCreatedEntitiesTestData();
            Object obj = CyclicXmlHandler.Read(testData.xmlJava);
            List<Object> expecteds = (List<Object>)testData.obj;
            List<Object> actuals = (List<Object>)obj;
            Assert.AssertEquals(expecteds.Count, actuals.Count);
            for (int i = 0; i < expecteds.Count; i++)
            {
                Material expected = expecteds[i] as Material;
                Material actual = actuals[i] as Material;
                AssertEquals(expected, actual);
            }
        }

        [TestMethod]
        public void ReadCreatedAndExistingChildren()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedAndExistingChildrenTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            AssertEquals((EntityA)testData.obj, (EntityA)obj);
        }

        [TestMethod]
        public void ReadCreatedAndExistingChildren2()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedAndExistingChildrenTestData2();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            AssertEquals((EntityA)testData.obj, (EntityA)obj);
        }

        [TestMethod]
        public void ReadCreatedAndExistingChildrenJavaXML()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedAndExistingChildrenTestData();
            Object obj = CyclicXmlHandler.Read(testData.xmlJava);
            AssertEquals((EntityA)testData.obj, (EntityA)obj);
        }

        [TestMethod]
        public void ReadUpdatedEntity()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getUpdatedEntityTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            Assert.AssertEquals(testData.obj, obj);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadUpdatedEntityJavaXML()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getUpdatedEntityTestData();
            Object obj = CyclicXmlHandler.Read(testData.xmlJava);
            Assert.AssertEquals(testData.obj, obj);
            AssertEquals((Material)testData.obj, (Material)obj);
        }

        [TestMethod]
        public void ReadCreatedAndUpdatedEntities()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedAndUpdatedEntitiesTestData();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            List<Object> expecteds = (List<Object>)testData.obj;
            List<Object> actuals = (List<Object>)obj;
            Assert.AssertEquals(expecteds.Count, actuals.Count);
            for (int i = 0; i < expecteds.Count; i++)
            {
                Material expected = expecteds[i] as Material;
                Material actual = actuals[i] as Material;
                AssertEquals(expected, actual);
            }
        }

        [TestMethod]
        public void ReadCreatedAndUpdatedEntities2()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedAndUpdatedEntitiesTestData2();
            Object obj = CyclicXmlHandler.Read(testData.xml);
            List<Object> expecteds = (List<Object>)testData.obj;
            List<Object> actuals = (List<Object>)obj;
            Assert.AssertEquals(expecteds.Count, actuals.Count);
            for (int i = 0; i < expecteds.Count; i++)
            {
                Material expected = expecteds[i] as Material;
                Material actual = actuals[i] as Material;
                AssertEquals(expected, actual);
            }
        }

        [TestMethod]
        public void ReadCreatedAndUpdatedEntitiesJavaXML()
        {
            InitManually(GetType());
            OriWrapperTestBed.TestData testData = OriWrapperTestBed.getCreatedAndUpdatedEntitiesTestData();
            Object obj = CyclicXmlHandler.Read(testData.xmlJava);
            List<Object> expecteds = (List<Object>)testData.obj;
            List<Object> actuals = (List<Object>)obj;
            Assert.AssertEquals(expecteds.Count, actuals.Count);
            for (int i = 0; i < expecteds.Count; i++)
            {
                Material expected = expecteds[i] as Material;
                Material actual = actuals[i] as Material;
                AssertEquals(expected, actual);
            }
        }

        private void AreArraysEqual(object[] expected, object[] actual)
        {
            Assert.AssertEquals(expected.Length, actual.Length);
            for (int i = 0; i < expected.Length; i++)
            {
                Assert.AssertEquals(expected[i], actual[i]);
            }
        }

        protected void AreCollectionsEqual(ICollection expected, ICollection actual)
        {
            Assert.AssertEquals(expected.Count, actual.Count);
            IEnumerator expectedIter = expected.GetEnumerator();
            IEnumerator actualIter = actual.GetEnumerator();
            int index = 0;
            while (expectedIter.MoveNext())
            {
                Object expectedEntry = expectedIter.Current;
                actualIter.MoveNext();
                Object actualEntry = actualIter.Current;
                Assert.AssertEquals(expectedEntry, actualEntry, "collection first differend at element " + index++);
            }
        }

        protected void AreSetsEqual(ISet<Object> expected, ISet<Object> actual)
        {
            Assert.AssertEquals(expected.Count, actual.Count);
            IEnumerator expectedIter = expected.GetEnumerator();
            while (expectedIter.MoveNext())
            {
                Object expectedEntry = expectedIter.Current;
                Assert.AssertTrue(actual.Contains(expectedEntry), "expected does not contain element " + expectedEntry);
            }
        }

        protected void AssertEquals(Material expected, Material actual)
        {
            Assert.AssertEquals(expected.Buid, actual.Buid);
            Assert.AssertEquals(expected.CreatedBy, actual.CreatedBy);
            Assert.AssertEquals(expected.CreatedOn, actual.CreatedOn);
            Assert.AssertEquals(expected.Id, actual.Id);
            Assert.AssertEquals(expected.Name, actual.Name);
            Assert.AssertEquals(expected.UpdatedBy, actual.UpdatedBy);
            Assert.AssertEquals(expected.UpdatedOn, actual.UpdatedOn);
            Assert.AssertEquals(expected.Version, actual.Version);

            AssertEquals(expected.MaterialGroup, actual.MaterialGroup);
        }

        protected void AssertEquals(MaterialGroup expected, MaterialGroup actual)
        {
            if (expected == null)
            {
                Assert.AssertNull(actual);
                return;
            }

            Assert.AssertNotNull(actual);
            Assert.AssertEquals(expected.Buid, actual.Buid);
            Assert.AssertEquals(expected.Id, actual.Id);
            Assert.AssertEquals(expected.Name, actual.Name);
            Assert.AssertEquals(expected.Version, actual.Version);
        }

        private void AssertEquals(EntityA expected, EntityA actual)
        {
            if (expected == null)
            {
                Assert.AssertNull(actual);
                return;
            }

            Assert.AssertNotNull(actual);
            Assert.AssertEquals(expected.Id, actual.Id);
            Assert.AssertEquals(expected.Version, actual.Version);
            Assert.AssertEquals(expected.EntityBs.Count, actual.EntityBs.Count);
            for (int i = 0; i < expected.EntityBs.Count; i++)
            {
                AssertEquals(expected.EntityBs[i], actual.EntityBs[i]);
            }
        }

        private void AssertEquals(EntityB expected, EntityB actual)
        {
            if (expected == null)
            {
                Assert.AssertNull(actual);
                return;
            }

            Assert.AssertNotNull(actual);
            Assert.AssertEquals(expected.Id, actual.Id);
            Assert.AssertEquals(expected.Version, actual.Version);
        }
    }
}

File: jambeth/jambeth-test/src/test/java/de/osthus/ambeth/xml/oriwrapper/OriWrapperReadTest.java
package de.osthus.ambeth.xml.oriwrapper;

import java.util.Collection;
import java.util.Iterator;
import java.util.List;

import org.junit.Assert;
import org.junit.Test;

import de.osthus.ambeth.config.ServiceConfigurationConstants;
import de.osthus.ambeth.ioc.BootstrapScannerModule;
import de.osthus.ambeth.ioc.XmlModule;
import de.osthus.ambeth.ioc.annotation.Autowired;
import de.osthus.ambeth.model.Material;
import de.osthus.ambeth.model.MaterialGroup;
import de.osthus.ambeth.service.ProcessServiceTestModule;
import de.osthus.ambeth.testutil.AbstractPersistenceTest;
import de.osthus.ambeth.testutil.SQLData;
import de.osthus.ambeth.testutil.SQLDataList;
import de.osthus.ambeth.testutil.SQLStructure;
import de.osthus.ambeth.testutil.SQLStructureList;
import de.osthus.ambeth.testutil.TestModule;
import de.osthus.ambeth.testutil.TestProperties;
import de.osthus.ambeth.testutil.TestPropertiesList;
import de.osthus.ambeth.transfer.ServiceDescription;
import de.osthus.ambeth.xml.ICyclicXMLHandler;
import de.osthus.ambeth.xml.oriwrapper.OriWrapperTestBed.TestData;

@SQLStructureList({ @SQLStructure("../../service/ProcessServiceTest_structure.sql"), @SQLStructure("OriWrapper_structure.sql") })
@SQLDataList({ @SQLData("../../service/ProcessServiceTest_data.sql"), @SQLData("OriWrapper_data.sql") })
@TestPropertiesList({ @TestProperties(file = "de/osthus/ambeth/xml/oriwrapper/OriWrapperTestData.properties"),
		@TestProperties(name = ServiceConfigurationConstants.mappingFile, value = "de/osthus/ambeth/service/orm.xml;de/osthus/ambeth/xml/oriwrapper/orm.xml") })
@TestModule({ BootstrapScannerModule.class, XmlModule.class, ProcessServiceTestModule.class, OriWrapperTestModule.class })
public class OriWrapperReadTest extends AbstractPersistenceTest
{
	@Autowired(XmlModule.CYCLIC_XML_HANDLER)
	protected ICyclicXMLHandler cyclicXmlHandler;

	@Autowired
	protected OriWrapperTestBed oriWrapperTestBed;

	@Override
	public void afterPropertiesSet() throws Throwable
	{
		super.afterPropertiesSet();

		oriWrapperTestBed.init();
	}

	@Test
	public void readSimpleEntity()
	{
		TestData testData = oriWrapperTestBed.getSimpleEntityTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertEquals(testData.obj, obj);
	}

	@Test
	public void readEntityWithRelation()
	{
		TestData testData = oriWrapperTestBed.getEntityWithRelationTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertEquals(testData.obj, obj);
	}

	@Test
	public void readMixedArray()
	{
		TestData testData = oriWrapperTestBed.getMixedArrayTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertArrayEquals((Object[]) testData.obj, (Object[]) obj);
	}

	@Test
	public void readMixedList()
	{
		TestData testData = oriWrapperTestBed.getMixedListTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		assertCollectionEquals((Collection<?>) testData.obj, (Collection<?>) obj);
	}

	@Test
	public void readMixedLinkedSet()
	{
		TestData testData = oriWrapperTestBed.getMixedLinkedSetTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		assertCollectionEquals((Collection<?>) testData.obj, (Collection<?>) obj, false);
	}

	@Test
	public void readServiceDescription() throws SecurityException, NoSuchMethodException
	{
		TestData testData = oriWrapperTestBed.getServiceDescriptionTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		de.osthus.ambeth.transfer.Assert.assertEquals((ServiceDescription) testData.obj, (ServiceDescription) obj);
	}

	@Test
	public void readCreatedEntity()
	{
		TestData testData = oriWrapperTestBed.getCreatedEntityTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readCreatedEntityCSXML()
	{
		TestData testData = oriWrapperTestBed.getCreatedEntityTestData();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readCreatedChildEntity()
	{
		TestData testData = oriWrapperTestBed.getCreatedChildEntityTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readCreatedChildEntity2()
	{
		TestData testData = oriWrapperTestBed.getCreatedChildEntityTestData2();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readCreatedChildEntityCSXML()
	{
		TestData testData = oriWrapperTestBed.getCreatedChildEntityTestData();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readCreatedChildEntityCSXML2()
	{
		TestData testData = oriWrapperTestBed.getCreatedChildEntityTestData2();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readCreatedParentAndChildEntities()
	{
		TestData testData = oriWrapperTestBed.getCreatedParentAndChildEntitiesTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readCreatedParentAndChildEntities2()
	{
		TestData testData = oriWrapperTestBed.getCreatedParentAndChildEntitiesTestData2();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readCreatedParentAndChildEntitiesCSXML()
	{
		TestData testData = oriWrapperTestBed.getCreatedParentAndChildEntitiesTestData();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readCreatedParentAndChildEntitiesCSXML2()
	{
		TestData testData = oriWrapperTestBed.getCreatedParentAndChildEntitiesTestData2();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		Assert.assertTrue(obj instanceof Material);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readCreatedParentAndChildEntitiesInList()
	{
		TestData testData = oriWrapperTestBed.getCreatedParentAndChildEntitiesInListTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertTrue(obj instanceof List);
		List<Object> actuals = (List<Object>) obj;
		assertEquals((Material) ((List<Object>) testData.obj).get(0), (Material) actuals.get(0));
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readCreatedParentAndChildEntitiesInListCSXML()
	{
		TestData testData = oriWrapperTestBed.getCreatedParentAndChildEntitiesInListTestData();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		Assert.assertTrue(obj instanceof List);
		List<Object> actuals = (List<Object>) obj;
		assertEquals((Material) ((List<Object>) testData.obj).get(0), (Material) actuals.get(0));
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readCreatedParentAndChildEntitiesInListCSXML2()
	{
		TestData testData = oriWrapperTestBed.getCreatedParentAndChildEntitiesInListTestData2();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		Assert.assertTrue(obj instanceof List);
		List<Object> actuals = (List<Object>) obj;
		assertEquals((Material) ((List<Object>) testData.obj).get(0), (Material) actuals.get(0));
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readCreatedParentAndChildEntitiesInList2()
	{
		TestData testData = oriWrapperTestBed.getCreatedParentAndChildEntitiesInListTestData2();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertTrue(obj instanceof List);
		List<Object> actuals = (List<Object>) obj;
		assertEquals((Material) ((List<Object>) testData.obj).get(0), (Material) actuals.get(0));
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readMultipleCreatedEntities()
	{
		TestData testData = oriWrapperTestBed.getMultipleCreatedEntitiesTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		List<Material> expecteds = (List<Material>) testData.obj;
		List<Material> actuals = (List<Material>) obj;
		Assert.assertEquals(expecteds.size(), actuals.size());
		for (int i = 0; i < expecteds.size(); i++)
		{
			Material expected = expecteds.get(i);
			Material actual = actuals.get(i);
			assertEquals(expected, actual);
		}
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readMultipleCreatedEntities2()
	{
		TestData testData = oriWrapperTestBed.getMultipleCreatedEntitiesTestData2();
		Object obj = cyclicXmlHandler.read(testData.xml);
		List<Material> expecteds = (List<Material>) testData.obj;
		List<Material> actuals = (List<Material>) obj;
		Assert.assertEquals(expecteds.size(), actuals.size());
		for (int i = 0; i < expecteds.size(); i++)
		{
			Material expected = expecteds.get(i);
			Material actual = actuals.get(i);
			assertEquals(expected, actual);
		}
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readMultipleCreatedEntitiesCSXML()
	{
		TestData testData = oriWrapperTestBed.getMultipleCreatedEntitiesTestData();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		List<Material> expecteds = (List<Material>) testData.obj;
		List<Material> actuals = (List<Material>) obj;
		Assert.assertEquals(expecteds.size(), actuals.size());
		for (int i = 0; i < expecteds.size(); i++)
		{
			Material expected = expecteds.get(i);
			Material actual = actuals.get(i);
			assertEquals(expected, actual);
		}
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readMultipleCreatedEntitiesCSXML2()
	{
		TestData testData = oriWrapperTestBed.getMultipleCreatedEntitiesTestData2();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		List<Material> expecteds = (List<Material>) testData.obj;
		List<Material> actuals = (List<Material>) obj;
		Assert.assertEquals(expecteds.size(), actuals.size());
		for (int i = 0; i < expecteds.size(); i++)
		{
			Material expected = expecteds.get(i);
			Material actual = actuals.get(i);
			assertEquals(expected, actual);
		}
	}

	@Test
	public void readUpdatedEntity()
	{
		TestData testData = oriWrapperTestBed.getUpdatedEntityTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertEquals(testData.obj, obj);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@Test
	public void readBuidUpdatedEntity()
	{
		TestData testData = oriWrapperTestBed.getBuidUpdatedEntityTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		Assert.assertEquals(testData.obj, obj);
		assertEquals((Material) testData.obj, (Material) obj);
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readCreatedAndUpdatedEntities()
	{
		TestData testData = oriWrapperTestBed.getCreatedAndUpdatedEntitiesTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		List<Material> expecteds = (List<Material>) testData.obj;
		List<Material> actuals = (List<Material>) obj;
		Assert.assertEquals(expecteds.size(), actuals.size());
		for (int i = 0; i < expecteds.size(); i++)
		{
			Material expected = expecteds.get(i);
			Material actual = actuals.get(i);
			assertEquals(expected, actual);
		}
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readCreatedAndUpdatedEntities2()
	{
		TestData testData = oriWrapperTestBed.getCreatedAndUpdatedEntitiesTestData2();
		Object obj = cyclicXmlHandler.read(testData.xml);
		List<Material> expecteds = (List<Material>) testData.obj;
		List<Material> actuals = (List<Material>) obj;
		Assert.assertEquals(expecteds.size(), actuals.size());
		for (int i = 0; i < expecteds.size(); i++)
		{
			Material expected = expecteds.get(i);
			Material actual = actuals.get(i);
			assertEquals(expected, actual);
		}
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readCreatedAndUpdatedEntitiesCSXML()
	{
		TestData testData = oriWrapperTestBed.getCreatedAndUpdatedEntitiesTestData();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		List<Material> expecteds = (List<Material>) testData.obj;
		List<Material> actuals = (List<Material>) obj;
		Assert.assertEquals(expecteds.size(), actuals.size());
		for (int i = 0; i < expecteds.size(); i++)
		{
			Material expected = expecteds.get(i);
			Material actual = actuals.get(i);
			assertEquals(expected, actual);
		}
	}

	@SuppressWarnings("unchecked")
	@Test
	public void readCreatedAndUpdatedEntitiesCSXML2()
	{
		TestData testData = oriWrapperTestBed.getCreatedAndUpdatedEntitiesTestData2();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		List<Material> expecteds = (List<Material>) testData.obj;
		List<Material> actuals = (List<Material>) obj;
		Assert.assertEquals(expecteds.size(), actuals.size());
		for (int i = 0; i < expecteds.size(); i++)
		{
			Material expected = expecteds.get(i);
			Material actual = actuals.get(i);
			assertEquals(expected, actual);
		}
	}

	@Test
	public void readCreatedAndExistingChildren()
	{
		TestData testData = oriWrapperTestBed.getCreatedAndExistingChildrenTestData();
		Object obj = cyclicXmlHandler.read(testData.xml);
		assertEquals((EntityA) testData.obj, (EntityA) obj);
	}

	@Test
	public void readCreatedAndExistingChildren2()
	{
		TestData testData = oriWrapperTestBed.getCreatedAndExistingChildrenTestData2();
		Object obj = cyclicXmlHandler.read(testData.xml);
		assertEquals((EntityA) testData.obj, (EntityA) obj);
	}

	@Test
	public void readCreatedAndExistingChildrenCSXML()
	{
		TestData testData = oriWrapperTestBed.getCreatedAndExistingChildrenTestData();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		assertEquals((EntityA) testData.obj, (EntityA) obj);
	}

	@Test
	public void readCreatedAndExistingChildrenCSXML2()
	{
		TestData testData = oriWrapperTestBed.getCreatedAndExistingChildrenTestData2();
		Object obj = cyclicXmlHandler.read(testData.xmlCS);
		assertEquals((EntityA) testData.obj, (EntityA) obj);
	}

	protected void assertCollectionEquals(Collection<?> expected, Collection<?> actual)
	{
		assertCollectionEquals(expected, actual, true);
	}

	protected void assertCollectionEquals(Collection<?> expected, Collection<?> actual, boolean inOrder)
	{
		Assert.assertEquals(expected.size(), actual.size());
		Iterator<?> expectedIter = expected.iterator();
		Iterator<?> actualIter = actual.iterator();
		int index = 0;
		while (expectedIter.hasNext())
		{
			Object expectedEntry = expectedIter.next();
			if (inOrder)
			{
				Object actualEntry = actualIter.next();
				Assert.assertEquals("collection first differend at element " + index++, expectedEntry, actualEntry);
			}
			else
			{
				Assert.assertTrue(actual.contains(expectedEntry));
			}
		}
	}

	private void assertEquals(Material expected, Material actual)
	{
		Assert.assertEquals(expected.getBuid(), actual.getBuid());
		Assert.assertEquals(expected.getCreatedBy(), actual.getCreatedBy());
		Assert.assertEquals(expected.getCreatedOn(), actual.getCreatedOn());
		Assert.assertEquals(expected.getId(), actual.getId());
		Assert.assertEquals(expected.getName(), actual.getName());
		Assert.assertEquals(expected.getUpdatedBy(), actual.getUpdatedBy());
		Assert.assertEquals(expected.getUpdatedOn(), actual.getUpdatedOn());
		Assert.assertEquals(expected.getVersion(), actual.getVersion());

		assertEquals(expected.getMaterialGroup(), actual.getMaterialGroup());
		Assert.assertEquals(expected.getUnit(), actual.getUnit());
	}

	private void assertEquals(MaterialGroup expected, MaterialGroup actual)
	{
		if (expected == null)
		{
			Assert.assertNull(actual);
			return;
		}

		Assert.assertNotNull(actual);
		Assert.assertEquals(expected.getBuid(), actual.getBuid());
		Assert.assertEquals(expected.getId(), actual.getId());
		Assert.assertEquals(expected.getName(), actual.getName());
		Assert.assertEquals(expected.getVersion(), actual.getVersion());
	}

	private void assertEquals(EntityA expected, EntityA actual)
	{
		if (expected == null)
		{
			Assert.assertNull(actual);
			return;
		}

		Assert.assertNotNull(actual);
		Assert.assertEquals(expected.getId(), actual.getId());
		Assert.assertEquals(expected.getVersion(), actual.getVersion());
		Assert.assertEquals(expected.getEntityBs().size(), actual.getEntityBs().size());
		for (int i = 0; i < expected.getEntityBs().size(); i++)
		{
			assertEquals(expected.getEntityBs().get(i), actual.getEntityBs().get(i));
		}
	}

	private void assertEquals(EntityB expected, EntityB actual)
	{
		if (expected == null)
		{
			Assert.assertNull(actual);
			return;
		}

		Assert.assertNotNull(actual);
		Assert.assertEquals(expected.getId(), actual.getId());
		Assert.assertEquals(expected.getVersion(), actual.getVersion());
	}
}
